File: legacy_code/stacks/min_stack.py
#!/bin/python
class MinStack(object):

    def __init__(self):
        self.stack = []
        self.min_stack = []


    def top(self):
        if self.stack:
            return self.stack[-1]
        else:
            raise Exception("Empty stack")


    def min(self):
        if self.min_stack:
            return self.min_stack[-1]
        else:
            raise Exception("Empty stack")


    def push(self, value):
        self.stack.append(value)
        if not self.min_stack or value <= self.min_stack[-1]:
            self.min_stack.append(value)


    def pop(self):
        value = self.stack.pop()
        if value <= self.min_stack[-1]:
            self.min_stack.pop()

File: legacy_code/data_structures/stack/min_stack.cpp
#include <iostream>
#include <stack>

using namespace std;

class MinStack {

    private:
        stack<int> s;
        stack<int> s_min;

    public:

        void push(int data) {
            s.push(data);
            if (s_min.empty() || data <= s_min.top()) {
                s_min.push(data);
            }
        }

        void pop() {
            int data = s.top();
            s.pop();
            if (data <= s_min.top()) {
                s_min.pop();
            }
        }

        int top() {
            return s.top();
        }

        int min() {
            return s_min.top();
        }
};

void push(MinStack* min_s, int item) {
    cout << "PUSHING: " << item << endl;
    min_s->push(item);
    cout << "TOP: "<< min_s->top() << " MIN: " << min_s->min() << endl << endl;
}

void pop(MinStack* min_s) {
    cout << "POPING: " << min_s->top() << endl;
    min_s->pop();
    cout << "TOP: "<< min_s->top() << " MIN: " << min_s->min() << endl << endl;
}

int main() {
    MinStack min_s;
    push(&min_s, 5);
    push(&min_s, 6);
    push(&min_s, 7);
    push(&min_s, 5);
    push(&min_s, 4);
    push(&min_s, 9);
    pop(&min_s);
    pop(&min_s);
    pop(&min_s);
    pop(&min_s);
    pop(&min_s);
    return 0;
}
